### ISSUE TITLE
TST RecursionError maximum recursion in test_zarr

### ISSUE THREAD

**Issue by rth** [pyodide/pyodide#1629]
`zarr/test_zarr.py` is currently consistently failing consistently in Chrome with a recursion error. 

It might be a good occasion to compile Pyodide with `-Os` for smaller stack frame size https://github.com/pyodide/pyodide/issues/1541 and increasing the recursion limit a bit.

**Comment by mgreminger** [comment]
@rth, I've found that the [heuristics](https://github.com/pyodide/pyodide/blob/b401350a2a972ad48a80d8def635c310fa9acbb0/src/js/pyodide.js#L140) that are currently used to set the recursion limit are overly conservative, especially in web workers in Firefox (maybe this is the case for the test_zarr failure as well). For the app I'm working on, I need to load SymPy. I found that SymPy needs a recursion limit of at least 131 to load successfully. If I let Pyodide set the recursion limit,  about 10% of the time, SymPy will fail to load because of too low a recursion limit being set by Pyodide (when using Firefox in a web worker). I now manually set the recursion limit to 200 and I've never had the app fail in Firefox or Chrome due to exceeding the stack limit. With the improved error handling in version 0.17 that raises a fatal error if the call stack is exceeded, I think it would be better to make the default recursion limit a fixed number (200 or even 400, I think 1000 is the default for cpython). We would probably want to change the fatal error message from the [current message](https://github.com/pyodide/pyodide/blob/b401350a2a972ad48a80d8def635c310fa9acbb0/src/js/pyodide.js#L64) to one that includes the possibility that the call stack was exceeded to make debugging a call stack error easier. I'm currently using a try-catch around my python calls that looks for RangeError (chrome) or InternalError (firefox) so that I know if the call stack is exceeded, but it has never exceeded the call stack in my app under quite a bit of testing.

**Comment by rth** [comment]
I agree with you that the current situation need to be improved. We have [many issues](https://github.com/pyodide/pyodide/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+recursion)  on this topic.

With the current heuristic on 0.17.0 I get,

| Browser                                | sys.getrecusionlimit() |
|----------------------------------------|------------------------|
| Firefox 89 (Linux)                     | 637                    |
| Firefox Mobile 89                      | 267                    |
| Chrome 89 (Linux), Chrome 91 (Windows) | 279                    |
| Safari 13.1 (Mac OS Catalina)          | 750                    |


With https://github.com/pyodide/pyodide/pull/1581 it would be a bit less (though now I'm not convinced we should have made that change).

So overall the issue is Chrome. We could very well try to set the minimum to 300 or 400 (unless it's larger with this heuristic), but then maybe there is still the question of how to deal with fatal stack overflow errors when they do happen https://github.com/pyodide/pyodide/issues/1605

**Comment by mgreminger** [comment]
For web workers, the heuristic gives about 150 for both Firefox 89 (Windows) and Chrome 91 (windows). About 1 in 10 times, Firefox goes below 131 (as low as 117), which no longer allows SymPy to load. 

I see your point about the fatal stack error being an issue, but, for some apps, hitting the python recursion limit is effectively a fatal error since it often happens when loading packages that are needed by the app. In these cases, setting the recursion limit higher gives the highest probability that the code will run successfully since we can't predict what the highest safe recursion limit will be.

**Comment by hoodmane** [comment]
@mgreminger One problem is that the fatal errors cause trouble in our test suite. It takes a long time for the runtime to start up and if there is a fatal error then it requires the runtime to be refreshed before it can be used again. Right now it is statically determined which collections of tests share the same Pyodide instance and after a fatal error all of the remaining tests run in the same instance all just fail. 

We could xfail all of the tests that trigger recursion errors. One thing that might be interesting would be to update the test suite to refresh the runtime if a fatal error is triggered...

**Comment by mgreminger** [comment]
Good point, there are definitely cases where a Python runtime crash needs to be avoided such as the testing example you give or in an interactive REPL. Maybe best approach would be to add something about recursion limit to the docs since I think this is a pain point for people who are just starting to use Pyodide. I'll look at trying to do a PR for a documentation update on the recursion limit topic (when it might make sense to override recursion limit in your own code and what the implications of that choice are, for example), will probably take me a couple of weeks to get to it though.

**Comment by hoodmane** [comment]
Thanks @mgreminger! That would be very much appreciated. It's true, it's not like this is a private feature, if you need a higher recursion limit, all you have to do is say e.g.
```js
pyodide.runPython("import sys; sys.setrecursionlimit(400)");
```

**Comment by hoodmane** [comment]
Closing as resolved by #2019 (if not by some earlier change).
